### PR TITLE
support sector class + dynamic PoSt and PoRep configuration

### PR DESCRIPTION
**Pull request by laser**
Blocks [this downstream go-filecoin PR](https://github.com/filecoin-project/go-filecoin/pull/2549).

## What's in this PR?

- Accept config values through FFI used to control PoRep partitions, PoSt partitions, and sector size
- Plumb `PoStConfig` and `PoRep` config into `internal.rs` functions to make choice of partitions and size dynamic

## Why does this PR exist?

- Initial push to allow FFI consumers to control proofs behavior ("sector class")
- Initial work to support multiple sector sizes

### CODE REVIEW

File: filecoin-proofs/src/api/internal.rs [hunk 1]
```diff
@@ -418,10 +427,13 @@ fn verify_post_fixed_sectors_count(
         faults: fixed.faults.clone(),
     };
 
-    let verifying_key = get_post_verifying_key(fixed.sector_bytes)?;
+    let verifying_key = get_post_verifying_key(fixed.post_config)?;
 
-    let proof =
-        MultiProof::new_from_reader(Some(POST_PARTITIONS), &fixed.proof[0..192], &verifying_key)?;
+    let proof = MultiProof::new_from_reader(
+        Some(usize::from(PoStProofPartitions::from(fixed.post_config))),
```
Comment (laser) [inline filecoin-proofs/src/api/internal.rs]: Note that PoSt proof partitions is now dynamic here.

File: filecoin-proofs/src/api/internal.rs [hunk 2]
```diff
@@ -503,9 +515,9 @@ pub fn seal<T: Into<PathBuf> + AsRef<Path>>(
     let replica_id = replica_id::<DefaultTreeHasher>(prover_id, sector_id);
 
     let compound_setup_params = compound_proof::SetupParams {
-        vanilla_params: &setup_params(sector_config.sector_bytes()),
+        vanilla_params: &setup_params(PaddedBytesAmount::from(porep_config)),
         engine_params: &(*ENGINE_PARAMS),
-        partitions: Some(POREP_PARTITIONS),
+        partitions: Some(usize::from(PoRepProofPartitions::from(porep_config))),
```
Comment (laser) [inline filecoin-proofs/src/api/internal.rs]: Note that PoRep proof partitions is now dynamic here.

File: filecoin-proofs/src/api/internal.rs [hunk 3]
```diff
@@ -546,10 +558,12 @@ pub fn seal<T: Into<PathBuf> + AsRef<Path>>(
         &groth_params,
     )?;
 
+    // TODO: POREP_PROOF_BYTES needs to be a function of PoRepPartitions
```
Comment (laser) [inline filecoin-proofs/src/api/internal.rs]: Note that PoRep proof bytes is still fixed, below.

File: filecoin-proofs/src/api/internal.rs [hunk 4]
```diff
@@ -546,10 +558,12 @@ pub fn seal<T: Into<PathBuf> + AsRef<Path>>(
...
     let mut buf = Vec::with_capacity(POREP_PROOF_BYTES);
 
     proof.write(&mut buf)?;
 
+    // TODO: POREP_PROOF_BYTES needs to be a function of PoRepPartitions
```
Comment (laser) [inline filecoin-proofs/src/api/internal.rs]: Note that PoRep proof bytes is also fixed on the next line.

File: filecoin-proofs/src/api/internal.rs [hunk 5]
```diff
@@ -158,29 +160,31 @@ fn get_post_params(
...
+    let public_params = public_params(PaddedBytesAmount::from(porep_config));
 
     let get_verifying_key =
         || ZigZagCompound::verifying_key(&public_params, &ENGINE_PARAMS).map_err(Into::into);
 
     Ok(lookup_verifying_key(
-        format!("ZIGZAG[{}]", usize::from(sector_bytes)),
+        format!(
+            "ZIGZAG[{}]",
+            usize::from(PaddedBytesAmount::from(porep_config))
```
Comment (laser) [inline filecoin-proofs/src/api/internal.rs]: Is it true that these keys will need to be updated to include PoRep partitions?

File: filecoin-proofs/src/api/internal.rs [hunk 6]
```diff
@@ -192,7 +196,10 @@ fn get_post_verifying_key(
     };
 
     Ok(lookup_verifying_key(
-        format!("POST[{}]", usize::from(sector_bytes)),
+        format!(
+            "POST[{}]",
```
Comment (laser) [inline filecoin-proofs/src/api/internal.rs]: Is it true that this key will need to be updated to include PoSt partitions?

File: filecoin-proofs/src/api/mod.rs [hunk 7]
```diff
@@ -561,3 +530,64 @@ pub unsafe extern "C" fn get_staged_sectors(
...
+    sector_size: FFISectorSize,
+    seal_proof_partitions: FFISealProofPartitions,
+    post_proof_partitions: FFIPoStProofPartitions,
+}
+
+#[allow(non_camel_case_types)]
+#[repr(C)]
+pub enum FFISectorSize {
+    SSB_OneKiB,
+    SSB_TwoHundredFiftySixMiB,
```
Comment (porcuquine) [inline filecoin-proofs/src/api/mod.rs]: I just want to check once more that we're comfortable with a model for change in which new sizes might be inserted between existing sizes, such that — for example — the underlying enum value `1` changes significance between releases. If these values go on chain anywhere, they will be more stably interpretable through time if we use the actual sizes. (I know we've discussed and you favor this approach, just calling this aspect out explicitly one last time.)

Comment (laser) [reply filecoin-proofs/src/api/mod.rs]: > might be inserted between existing sizes

Yep, I'm with you. The position in the enum isn't meaningful in either go-filecoin or rust-fil-proofs.

File: filecoin-proofs/src/api/internal.rs (hunk 5 above)
Comment (porcuquine) [reply filecoin-proofs/src/api/internal.rs]: Yes, groth parameters will need to be fetched by both sector size and number of partitions.

File: filecoin-proofs/src/api/internal.rs (hunk 6 above)
Comment (porcuquine) [reply filecoin-proofs/src/api/internal.rs]: Yes, verifying keys will need to be fetched by both sector size and number of partitions.

**Pull request merged**